File: src/java/org/objectweb/asm/commons/cfg/graph/Digraph.java
package org.objectweb.asm.commons.cfg.graph;

import java.util.*;

/**
 * @author Tyler Sedlar
 */
public class Digraph<V, E> implements Iterable<V> {

    private final Map<V, Set<E>> graph = new HashMap<>();
    private CyclomaticComplexity cyclomatic = null;

    @SuppressWarnings("unchecked")
    public Set<E> edgeAt(int index) {
        return (Set<E>) graph.values().toArray()[index];
    }

    public int size() {
        return graph.size();
    }

    public boolean containsVertex(V vertex) {
        return graph.containsKey(vertex);
    }

    public boolean containsEdge(V vertex, E edge) {
        return graph.containsKey(vertex) && graph.get(vertex).contains(edge);
    }

    public boolean addVertex(V vertex) {
        if (graph.containsKey(vertex)) {
            return false;
        }
        graph.put(vertex, new HashSet<E>());
        return true;
    }

    public void addEdge(V start, E dest) {
        if (!graph.containsKey(start)) {
            return;
        }
        graph.get(start).add(dest);
    }

    public void removeEdge(V start, E dest) {
        if (!graph.containsKey(start)) {
            return;
        }
        graph.get(start).remove(dest);
    }

    public Set<E> edgesFrom(V node) {
        return Collections.unmodifiableSet(graph.get(node));
    }

    public void graph(Digraph<V, E> graph) {
        this.graph.putAll(graph.graph);
    }

    @Override
    public final Iterator<V> iterator() {
        return graph.keySet().iterator();
    }

    @Override
    public String toString() {
        StringBuilder builder = new StringBuilder();
        for (V v : graph.keySet()) {
            builder.append("\n    ").append(v).append(" -> ").append(graph.get(v));
        }
        return builder.toString();
    }

    public void flush() {
        graph.clear();
    }

    /**
     * Gets the complexity for this graph.
     *
     * @param cached <t>true</t> to get the cached complexity.
     * @return the complexity for this graph.
     */
    public CyclomaticComplexity cyclomatic(boolean cached) {
        if (cached && cyclomatic != null) {
            return cyclomatic;
        }
        int edges = 0;
        for (Set<E> set : graph.values()) {
            edges += set.size();
        }
        int connections = 0;
        for (Set<E> set : graph.values()) {
            if (!set.isEmpty()) {
                connections++;
            }
        }
        return (cyclomatic = new CyclomaticComplexity(edges, size(), connections));
    }

    /**
     * Gets the complexity for this graph.
     *
     * @return the complexity for this graph.
     */
    public CyclomaticComplexity cyclomatic() {
        return cyclomatic(true);
    }
}

File: src/kotlin/kt/osrs/analysis/tree/flow/graph/Digraph.kt
package kt.osrs.analysis.tree.flow.graph

import java.util.*

/**
 * @author Tyler Sedlar
 */
open class Digraph<V, E> : Iterable<V> {

    private val graph = mutableMapOf<V, MutableSet<E>>()
    private var cyclomatic: CyclomaticComplexity? = null

    fun edgeAt(index: Int): MutableSet<E> {
        return graph.values.toTypedArray()[index]
    }

    fun size(): Int {
        return graph.size
    }

    fun containsVertex(vertex: V): Boolean {
        return graph.containsKey(vertex)
    }

    fun containsEdge(vertex: V, edge: E): Boolean {
        return graph.containsKey(vertex) && graph[vertex]!!.contains(edge)
    }

    fun addVertex(vertex: V): Boolean {
        if (graph.containsKey(vertex)) {
            return false
        }
        graph[vertex] = HashSet()
        return true
    }

    fun addEdge(start: V, dest: E) {
        if (!graph.containsKey(start)) {
            return
        }
        graph[start]!!.add(dest)
    }

    fun removeEdge(start: V, dest: E) {
        if (!graph.containsKey(start)) {
            return
        }
        graph[start]!!.remove(dest)
    }

    fun edgesFrom(node: V): Set<E> {
        return Collections.unmodifiableSet(graph[node])
    }

    fun consume(graph: Digraph<V, E>) {
        this.graph.putAll(graph.graph)
    }

    override fun iterator(): Iterator<V> {
        return graph.keys.iterator()
    }

    override fun toString(): String {
        val builder = StringBuilder()
        for (v in graph.keys) {
            builder.append("\n    ").append(v).append(" -> ").append(graph[v])
        }
        return builder.toString()
    }

    fun flush() {
        graph.clear()
    }

    /**
     * Gets the complexity for this consume.
     *
     * @param cached <t>true</t> to get the cached complexity.
     * @return the complexity for this consume.
     */
    @JvmOverloads
    fun cyclomatic(cached: Boolean = true): CyclomaticComplexity {
        if (cached && cyclomatic != null) {
            return cyclomatic!!
        }
        var edges = 0
        for (set in graph.values) {
            edges += set.size
        }
        var connections = 0
        for (set in graph.values) {
            if (!set.isEmpty()) {
                connections++
            }
        }
        cyclomatic = CyclomaticComplexity(edges, size(), connections)
        return cyclomatic!!
    }
}
/**
 * Gets the complexity for this consume.
 *
 * @return the complexity for this consume.
 */